src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -5594,3 +5600,47 @@ ExtractParametersFromParamList(ParamListInfo paramListInfo,
 																   parameterData->value);
 	}
 }
+
+
+/*
+ * Delegated functions which are not part of another transaction are naturally safe,
+ * but if it's part of an outer transaction (a 2PC), we should ensure that the current
+ * task will not do remote(connection) activity to ensure the transactional integrity
+ * of the current participating 2PC.
+ */
+static void
+EnsureDelegatedFunctionSafeExecution(List *taskList)
+{
+	if (taskList == NIL)
+	{
+		/* we allow executions with an empty task list (e.g. WHERE false) */
+		return;
+	}
+
+	if (!IsCitusInitiatedRemoteBackend())
+	{
+		/* application-initiated backends coordinate the 2PC, need no restrictions */
+		return;
+	}
+
+	if (!MyBackendIsInDisributedTransaction())
```

Comment:
We should not have this check / escape hatch either. On the workers, we only consider ourselves to be in a distributed transaction if the coordinator does a BEGIN. However, we do not send BEGIN for single-statement, single-shard queries, yet we could execute a function/trigger as a result of e.g. a single-row insert.

Reply:
Yes, we both are on the same page on our understanding. The fundamental premise of your comment, Citus-Initiated backends should *never* do remote tasks. I'm fine with that limitation, but it might break the legacy behavior of the customer applications. Please see the below comment for the scenario.